### PR TITLE
Make beam type 3 compatible with failure models

### PR DESCRIPTION
**Pull request by MaciekWronski**
#### Description of the feature or the bug

It allows using failure models for materials compatibles with beams type 3. For instance /FAIL/JOHNSON is available

#### Description of the changes

I have added failure buffer for storage of /FAIL model variables in global element buffer, necessary for beams type 3.

Main engine routines for the beams were modified, necessary failure modules were added  and johnson-cook failure model was  developped

### CODE REVIEW

File: engine/source/materials/fail/johnson_cook/fail_johnson_b.F [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      my_real                     ,INTENT(IN)    :: TIME    ! current time
+      my_real ,DIMENSION(NUPARAM) ,INTENT(IN)    :: UPARAM  ! failure model parameter array
+      my_real ,DIMENSION(NEL)     ,INTENT(IN)    :: DPLA    ! plastic strain increment
+      my_real ,DIMENSION(NEL)     ,INTENT(IN)    :: SVM     ! Von Mises stress
+      my_real ,DIMENSION(NEL)     ,INTENT(IN)    :: PRESSURE! element pressure
+      my_real ,DIMENSION(NEL)     ,INTENT(IN)    :: EPSD    ! strain rate
+      my_real ,DIMENSION(NEL)     ,INTENT(IN)    :: TSTAR   ! normalized temperature
+      my_real ,DIMENSION(NEL)     ,INTENT(INOUT) :: OFF     ! element desactivation flag
+      my_real ,DIMENSION(NEL)     ,INTENT(INOUT) :: DFMAX   ! maximum damage
+      my_real ,DIMENSION(NEL)     ,INTENT(OUT)   :: TDEL    ! desactivation time
```
Comment (laurent-altr) [inline engine/source/materials/fail/johnson_cook/fail_johnson_b.F:61]: Hello @MaciekWronski, 

Could you double-check the intent of TDEL? 
TDEL(i) is initialized only for some i. With INTENT(OUT), the other values may be replaced by garbage even they were initialized elsewhere.

File: engine/source/materials/fail/johnson_cook/fail_johnson_ib.F [hunk 2]
```diff
@@ -53,37 +53,37 @@ SUBROUTINE FAIL_JOHNSON_PI(
...
+      my_real ,DIMENSION(NEL)     ,INTENT(IN)    :: EPSD    ! strain rate
       my_real ,DIMENSION(NEL)     ,INTENT(IN)    :: TSTAR   ! normalized temperature
       my_real ,DIMENSION(NEL)     ,INTENT(IN)    :: OFF     ! element desactivation flag
       my_real ,DIMENSION(NEL)     ,INTENT(IN)    :: SIGNXX  ! stress component
       my_real ,DIMENSION(NEL)     ,INTENT(IN)    :: SIGNXY  ! stress component
       my_real ,DIMENSION(NEL)     ,INTENT(IN)    :: SIGNZX  ! stress component
       INTEGER ,DIMENSION(NEL)     ,INTENT(INOUT) :: FOFF    ! integration point desactivation flag
       my_real ,DIMENSION(NEL)     ,INTENT(INOUT) :: DFMAX   ! maximum damage
-      my_real ,DIMENSION(NEL)     ,INTENT(OUT)   :: TDEL    ! desactiovation time
+      my_real ,DIMENSION(NEL)     ,INTENT(OUT)   :: TDEL    ! desactivation time
```
Comment (laurent-altr) [inline engine/source/materials/fail/johnson_cook/fail_johnson_ib.F:64]: Same remark, 
TDEL may not be fully initialized after the call of this routine

Comment (MaciekWronski) [reply engine/source/materials/fail/johnson_cook/fail_johnson_ib.F]: I have changed both commits, it's safe now. thanks

**Comment by laurent-altr** [comment]
Hello @MaciekWronski, 

There are 4 commits in your PR, including a merge commit from main into beam_3: you may want to discard the merge commit, and squash the other ones.

Thank you, 
Laurent

**Review by laurent-altr — CHANGES_REQUESTED** [review]
Multiple commits

**Pull request closed without merging**